codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/EnumGenerator.kt

```diff
@@ -6,32 +6,92 @@
 import software.amazon.smithy.model.traits.EnumDefinition
 import software.amazon.smithy.model.traits.EnumTrait
 import software.amazon.smithy.rust.codegen.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.rustlang.docs
+import software.amazon.smithy.rust.codegen.rustlang.documentShape
 import software.amazon.smithy.rust.codegen.rustlang.rust
 import software.amazon.smithy.rust.codegen.rustlang.rustBlock
 import software.amazon.smithy.rust.codegen.rustlang.rustTemplate
 import software.amazon.smithy.rust.codegen.rustlang.withBlock
 import software.amazon.smithy.rust.codegen.smithy.RuntimeType
 import software.amazon.smithy.rust.codegen.smithy.expectRustMetadata
 import software.amazon.smithy.rust.codegen.util.doubleQuote
+import software.amazon.smithy.rust.codegen.util.getTrait
+import software.amazon.smithy.rust.codegen.util.orNull
 import software.amazon.smithy.rust.codegen.util.toPascalCase
-import java.lang.IllegalStateException
+
+/** Model that wraps [EnumDefinition] to calculate and cache values required to generate the Rust enum source. */
+internal class EnumMemberModel(private val definition: EnumDefinition) {
+    // Because enum variants always start with an upper case letter, they will never
+    // conflict with reserved words (which are always lower case), therefore, we never need
+    // to fall back to raw identifiers
+    private val unescapedName: String? = definition.name.orNull()?.toPascalCase()
+
+    val collidesWithUnknown: Boolean = unescapedName == EnumGenerator.UnknownVariant
+
+    /** Enum name with correct case format and collision resolution */
+    fun derivedName(): String = when (collidesWithUnknown) {
+        // If there is a variant named "Unknown", then rename it to "UnknownValue" so that it
+        // doesn't conflict with the code generator's "Unknown" variant that exists for backwards compatibility.
+        true -> "UnknownValue"
+        else -> checkNotNull(unescapedName) { "Enum variants must be named to derive a name. This is a bug." }
+    }
+
+    val value: String get() = definition.value
+
+    private fun renderDocumentation(writer: RustWriter) {
+        writer.docWithNote(
+            definition.documentation.orNull(),
+            when (collidesWithUnknown) {
+                true ->
+                    "`::${EnumGenerator.UnknownVariant}` has been renamed to `::${EnumGenerator.EscapedUnknownVariant}`. " +
+                        "`::${EnumGenerator.UnknownVariant}` refers to additional values that may have been added since " +
+                        "this enum was generated."
+                else -> null
+            }
+        )
+    }
+
+    fun render(writer: RustWriter) {
+        renderDocumentation(writer)
+        writer.write("${derivedName()},")
+    }
+}
+
+private fun RustWriter.docWithNote(doc: String?, note: String?) {
+    doc?.also { docs(it) }
+    note?.also {
+        doc?.also { write("///") }
```

Comment:
can you add a comment saying the point of this is to create a line break? I originally thought these were the `///` before `**Note**`

Reply:
For sure.